### DISCUSSION TITLE
wrap angles modulus 2 Pi

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#21778] [Q&A General] [answered]
Is there a simple way to wrap azimuthal angles in cylindrical coordinates back to [0, 2 Pi)? I guess using the C++ modulus function could be fraught as the wrapping should ideally be made during the executioner?

**Comment by mangerij** [comment]
I should add that we are interested in spherical coordinates. I can wrap the azimuthal angle in an UO (similar to the one @dschwen provided to renormalize the vector solution to be on the unit sphere).

To wrap, I just use this line:

```
for (const auto i : index_range(_var_numbers)) 
            if (solution(dof_indices[i][j]) > 2.0*libMesh::pi)
            solution.set(dof_indices[i][j], solution(dof_indices[i][j]) - 2.0*libMesh::pi * floor( solution(dof_indices[i][j]) / (2.0*libMesh::pi) ));
```

`j` here could mean more azimuthal angles - very useful

  Now we are trying (quite obviously because it gives a lot of speed up) to do the problem with just two degrees of freedom, the azimuthal [0,2pi) and polar angle [0,pi). However, subtracting modulo pi from the solution is not the same because it is a 180 degree rotation not 360. Any thoughts here?

**Comment by GiudGiud** [comment]
I m not sure I understand what you want to do here. 
In what forms are those angles? Are they in a variable like @mangerij is accessing? Is it an aux or nonlinear?

**Comment by heinono1** [comment]
Hi Guillaume,

They are the non-linear variables (spherical coordinates) and it is vital
to have them wrapped back appropriately to [0,2 Pi) (azimuth) and a little
more complicated wrapping for the polar angle to [0,Pi].

-Olle

On Thu, Aug 4, 2022 at 7:54 PM Guillaume Giudicelli <
***@***.***> wrote:

> I m not sure I understand what you want to do here.
> In what forms are those angles? Are they in a variable like @mangerij
> <https://github.com/mangerij> is accessing? Is it an aux or nonlinear?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21778#discussioncomment-3329560>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF7DXNZTJBAMHAHSHYLVXRQ4PANCNFSM55S7EUVA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by GiudGiud** [reply]
so they are the result of a non-linear solve?
We typically don't modify those after the solve, though it s totally possible with user objects

**Comment by heinono1** [comment]
Yes, they are the result of a non-linear solve. Ideally (just like with the
renormalization), one would like to do the wrapping at each step that
modifies the angles. The problem is that when the polar angle is close to 0
or Pi, there can be divergencies and/or discontinuities that makes the
solution diverge or crash.

On Thu, Aug 4, 2022 at 10:01 PM Guillaume Giudicelli <
***@***.***> wrote:

> so they are the result of a non-linear solve?
> We typically don't modify those after the solve, though it s totally
> possible with user objects
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/21778#discussioncomment-3330013>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEFZKSRFJT3MGOFAWMNLVXR7XPANCNFSM55S7EUVA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by GiudGiud** [reply]
I would go the user object route. You may execute user objects very often, as often as every linear iteration, so that can be helpful in the solve if the normalization is constantly needed

**Reply by GiudGiud** [reply]
Another thing you can try is leveraging the Bounds system to keep the nonlinear variable in the desired range. This could hinder the convergence of the solver but it might be worth a try
https://mooseframework.inl.gov/source/auxkernels/ConstantBoundsAux.html

**Comment by mangerij** [comment, marked as answer]
Hi,

I tried the Bounds and Damper systems and the solver crapped out at some point. At this moment, the wrapping is successfully done in a UO with

azimuthal:

` solution.set(dof_indices[i][j], solution(dof_indices[i][j]) - 2.0*libMesh::pi * floor( solution(dof_indices[i][j]) / (2.0*libMesh::pi) ));`
 
polar:

```
            if (floor( solution(dof_indices[i][j]) / (libMesh::pi) ) >= 1.0)
            {
              std::cout << "Found dof above polar angle limit, wrapping..." << std::endl;
              solution.set(dof_indices[i][j], 2.0*libMesh::pi - solution(dof_indices[i][j]) );
            }
            else if (floor( solution(dof_indices[i][j]) / (libMesh::pi) ) <= -1.0)
            {
              std::cout << "Found dof below polar angle limit, wrapping..." << std::endl;
              solution.set(dof_indices[i][j], -1.0*solution(dof_indices[i][j]) );
            }
```

but now we think we have something else unrelated that is problematic with our sph coord formulation. But thanks for the comments!
